src/backend/cuda/jit.cpp

```diff
@@ -69,54 +70,72 @@ struct Param {
 
     static const char *kernelVoid = "extern \"C\" __global__ void\n";
     static const char *dimParams =
-        "uint blocks_x, uint blocks_y, uint blocks_x_total, uint num_odims";
-
-    static const char *loopStart = R"JIT(
-    for (int blockIdx_x = blockIdx.x; blockIdx_x < blocks_x_total; blockIdx_x += gridDim.x) {
+        "int inc0, int inc1, int inc2, int inc3, "
+        "char decode0, char decode1, char decode2, char decode3";
+
+    static const char *blockStart = "{\n";
+    static const char *blockEnd   = "\n}\n\n";
+
+    static const char *linearIndexStart = R"JIT(
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;   // only dim0 since linear
+    if (idx < (int)outref.dims[2] * (int)outref.strides[2]) {
+        const int ostrides3 = outref.strides[3];
+        const int idxEnd = idx + (int)outref.dims[3] * ostrides3;
```

Comment:
The linear kernel has dimensions similar to [N,1,1,L] with
- N*L=elements in array
- N elements provided by blockIdx/threadIdx (only 1st 3 dimensions)
Remember: here the dimensions are directly (dim0,dim1,dim2) provided to CUDA/OPENCL and NOT (dim0*dim2,dim1*dim3,1) as is in the other kernels.

So the idx varies from [(0)..(N-1+overshoot)].  N=outref.strides[3].  idx is protected to [(0)..(N-1)]

In the example of [60, 1, 1, 4]:
- corresponding strides: [1, 60, 60, 60]
- dims[3]*strides[3] = 4 * 60 = 240
- ostrides3 = 60
- idx is protected between [0..59]
- idxEnd = idx + 4 * 60
- Now with 4x looping:
      - when idx starts at 0:    idxEnd=240; idx = [   0,   60, 120, 180], 240 excluded
      - when idx starts at 1:    idxEnd=241; idx = [   1,   61, 121, 181], 241 excluded
      - when idx starts at 59:  idxEnd=299; idx = [ 59, 119, 179, 239], 299 excluded

I do see your point, and idxEnd could be the same number (240) for all threads.
I will update the following:
in lineraIdxStart:  `idxEnd = ostrides3+dims[3] `
in linearIndexEnd: `while (idx < idxEnd)`

Reply:
After testing for all conditions, I had to return to the old situation.

Because I am not aware of the sign of the strides value, I can not use a greater or lower comparison.  The != is the only remaining comparison and I have to calculate based on the real iterator values (so idx + ...).  I added special comments to make this clear to future reviewers.